documented-derive/src/lib.rs

```diff
@@ -1,59 +1,130 @@
 use proc_macro::TokenStream;
 use quote::quote;
-use syn::{parse_macro_input, spanned::Spanned, DeriveInput, Error, Expr, ExprLit, Lit, Meta};
+use syn::{
+    parse_macro_input, parse_quote, spanned::Spanned, Attribute, Data, DataEnum, DataStruct,
+    DataUnion, DeriveInput, Error, Expr, ExprLit, Lit, Meta, Path,
+};
+
+fn crate_module_path() -> Path {
+    parse_quote!(::documented)
+}
 
 /// Derive proc-macro for `Documented` trait.
 #[proc_macro_derive(Documented)]
 pub fn documented(input: TokenStream) -> TokenStream {
     let input = parse_macro_input!(input as DeriveInput);
 
+    let doc_comments = match get_comment(&input.attrs) {
+        Ok(Some(doc)) => doc,
+        Ok(None) => {
+            return Error::new(input.ident.span(), "Missing doc comments")
+                .into_compile_error()
+                .into()
+        }
+        Err(e) => return e.into_compile_error().into(),
+    };
+
     let ident = input.ident;
 
-    let doc_comments = {
-        let maybe_str_literals = input
-            .attrs
-            .into_iter()
-            .filter_map(|attr| match attr.meta {
-                Meta::NameValue(name_value) if name_value.path.is_ident("doc") => {
-                    Some(name_value.value)
-                }
-                _ => None,
-            })
-            .map(|expr| match expr {
-                Expr::Lit(ExprLit { lit: Lit::Str(s), .. }) => Ok(s.value()),
-                e => Err(e),
-            })
-            .collect::<Result<Vec<_>, _>>();
-
-        let literals = match maybe_str_literals {
-            Ok(lits) => lits,
-            Err(expr) => {
-                return Error::new(expr.span(), "Doc comment is not a string literal")
-                    .into_compile_error()
-                    .into()
+    quote! {
+        #[automatically_derived]
+        impl documented::Documented for #ident {
+            const DOCS: &'static str = #doc_comments;
+        }
+    }
+    .into()
+}
+
+/// Derive proc-macro for `DocumentedFields` trait.
+#[proc_macro_derive(DocumentedFields)]
+pub fn documented_fields(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+
+    let fields_doc_comments = {
+        let fields_attrs: Vec<(syn::Ident, Vec<Attribute>)> = match input.data.clone() {
+            Data::Enum(DataEnum { variants, .. }) => {
+                variants.into_iter().map(|v| (v.ident, v.attrs)).collect()
             }
+            Data::Struct(DataStruct { fields, .. }) => fields
+                .into_iter()
+                .filter(|f| f.ident.is_some())
+                .map(|f| (f.ident.unwrap(), f.attrs))
+                .collect(),
```

Comment:
This is perhaps a bit too simplistic. The following is valid Rust after all:

```rust
struct Foo(
    /// Bar
    i32,
    /// Baz
    i32,
);
```

I think the better solution is to map tuple struct members to `0`, `1`, `2`, ... instead.

Reply:
Solved